File: python/pyplanes/FEM.py
#! /usr/bin/env python
# -*- coding:utf8 -*-
#
# FEM.py
#
# This file is part of cpplanes, a software distributed under the MIT license.
# For any question, please contact one of the authors cited below.
#
# Copyright (c) 2017
# 	Olivier Dazel <olivier.dazel@univ-lemans.fr>
# 	Mathieu Gaborit <gaborit@kth.se>
# 	Peter Göransson <pege@kth.se>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#

import numpy as np
import scipy.sparse as sparse
import scipy.sparse.linalg as splinalg


class FEM:
    """ Finite Element procedure

    I tried to keep it applicable to most geometries

    To do:
    - Make this class inherit from a Method super class for easier handling in C++
    - May be use multiple inheritance to inherit from Domain (which is finally a simple
      storage)
    - Method to Method hybrids have to be thought though... would it be easier to iterate
    """

    def __init__(self, interp, *args, **kwargs):
        """ Store domain and interpolation technique """
        super(FEM, self).__init__(*args, **kwargs)
        self.interp = interp

        self.__assembled_frequency = None

        self.map_dof_node = list(range(len(self.mesh.nodes)))
        self.removed_nodes = []
        self.A = None
        self.b = None

    def assemble(self, f=1000):

        k = self.material.get_wavenumber(f)

        self.__assembled_frequency = f
        self.__assembled_wavenumber = k

        A_r, A_c, A_v = [], [], []
        b_r, b_c, b_v = [], [], []

        for element in (self.mesh.get_element(_) for _ in range(len(self.mesh.elements))):

            (Qe, He) = self.interp.interpolate_over(element)
            elem_matrix = He-k**2*Qe

            for i_n1, n1 in enumerate(element.node_ids):
                for i_n2, n2 in enumerate(element.node_ids):
                    A_r.append(self.map_dof_node[n1])
                    A_c.append(self.map_dof_node[n2])
                    A_v.append(elem_matrix[i_n1, i_n2])

        for bc in self.boundary_conditions:
            bc.evaluate(f=f)
            A_r += bc.A_r
            A_c += bc.A_c
            A_v += bc.A_v
            b_r += bc.b_r
            b_c += bc.b_c
            b_v += bc.b_v
            self.removed_nodes += bc.null_nodes

        for n in self.removed_nodes:
            self.map_dof_node.remove(n)

        self.A = sparse.coo_matrix((A_v, (A_r, A_c)), dtype=np.complex128).tocsr()
        self.b = sparse.coo_matrix((b_v, (b_r, b_c)), shape=(self.A.shape[0], 1), dtype=np.complex128)
        self.b = self.b.toarray()

        deleted_rows = self.removed_nodes
        rows_mask = np.ones(self.A.shape[0], dtype=bool)
        rows_mask[deleted_rows] = False
        self.A = self.A[rows_mask]

        deleted_cols = self.removed_nodes
        self.A = self.A.tocsc()
        all_cols = np.arange(self.A.shape[1])
        cols_to_keep = np.where(np.logical_not(np.in1d(all_cols, deleted_cols)))[0]
        self.A = self.A[:, cols_to_keep]

        self.b = self.b[rows_mask]

    def solve(self, f=1000):
        if self.__assembled_frequency != f:
            self.assemble(f=f)
        if self.A is None or self.b is None:
            raise ValueError('System flagged as assembled but no value found.')

        self.x = splinalg.spsolve(self.A, self.b)

        self.solution = np.zeros((len(self.mesh.nodes),), dtype=np.complex128)
        self.solution[np.r_[self.map_dof_node]] = self.x
        return self.solution

File: src/FEM/FEM.cpp
/*
 * FEM.cpp
 *
 * This file is part of cpplanes, a software distributed under the MIT license.
 * For any question, please contact one of the authors cited below.
 *
 * Copyright (c) 2017
 * 	Olivier Dazel <olivier.dazel@univ-lemans.fr>
 * 	Mathieu Gaborit <gaborit@kth.se>
 * 	Peter Göransson <pege@kth.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 */

#include <numeric>

#include "FEM.h"

using namespace Eigen;

namespace cpplanes {

	template <typename _MeshT, class _Interp>
	FEM<_MeshT, _Interp>::FEM(
		const PhysicalModel& medium,
		const _MeshT& mesh,
		const _Interp& interp,
		const std::vector<std::shared_ptr<BCBase>> boundary_conditions
	):
		medium{medium}, mesh{mesh},
		interp{interp}, boundary_conditions{boundary_conditions} {

		map_dof_node.resize(mesh.element.size()*ElementT::nb_nodes);
		std::iota(map_dof_node.begin(), map_dof_node.end(), 0);

	}

	template <typename _MeshT, class _Interp>
	void FEM<_MeshT, _Interp>::assemble(cplx_t f) {
		cplx_t k = medium.get_Pwn(f);

		// Empty and resize the triplets' vector
		A_triplets.erase(A_triplets.begin(), A_triplets.end());
		A_triplets.resize(mesh.config.nodes_per_element^2*mesh.get_nb_elements());

		// Main assembly
		int list_position = 0;
		ElementT element;
		MatrixT Qe, He, elem_matrix;
		int i_n0, i_n1, n0, n1;
		for (int element_id = 0; element_id < mesh.elements.size(); ++element_id) {
			element = mesh.get_element(element_id);

			interp.get_Q(element, Qe);
			interp.get_H(element, He);

			elem_matrix = He-k^2*Qe;

			for (i_n0 = 0; i_n0 < ElementT::nb_nodes; ++i_n0) {
				for (i_n1 = 0; i_n1 < ElementT::nb_nodes; ++i_n1) {
					n0 = element.nodes_id(i_n0);
					n1 = element.nodes_id(i_n1);

					A_triplets.at(list_position) = {{
							map_dof_node.at(n0),
							map_dof_node.at(n1),
							elem_matrix(i_n0, i_n1)
						}
					};
				}
			}
			++list_position;
		}

		// Boundary conditions
		int nb_added_coefs = 0;
		for (auto bc : boundary_conditions ) {
			bc->evaluate(f);
			nb_added_coefs += bc->nb_added_coefs;
		}

		A_triplets.resize(A_triplets.size()+nb_added_coefs);
		b.resize(mesh.get_nb_elements()*ElementT::nb_nodes);

		// TODO: manage resizes
		for (auto bc : boundary_conditions ) {
			for (auto a_mod : bc->A) {
				A_triplets.at(list_position) = {{
						map_dof_node.at(int(a_mod.row())),
						map_dof_node.at(int(a_mod.col())),
						a_mod.value()
					}
				};
			}
			for (auto b_mod : bc->b) {
				b(map_dof_node.at(int(b_mod.at(0)))) += b_mod.at(1);
			}
		}

		A.resize(mesh.get_nb_elements()*ElementT::nb_nodes);
		A.setFromTriplets(A_triplets.begin(), A_triplets.end());


		assembled_frequency = f;
	}


	template <typename _MeshT, class _Interp>
	void FEM<_MeshT, _Interp>::solve(cplx_t f) {
		if (assembled_frequency != f) {
			assemble(f);
		}

		SparseLU<SparseMatrix<cplx_t, ColMajor>, COLAMDOrdering<Index> > solver;
		solver.analyzePattern(A);
		solver.factorize(A);

		x = solver.solve(b);
	}

}

